When creating S3-backed keyspace, check the endpoint instantly

xemul opened the pull request:
Currently CREATE KEYSPACE ... WITH STORAGE = { 'type' = 'S3' ... } will create keyspace even if the backend configuration is "invalid" in the sense that the requested endpoint is not known to scylla via object_storage.yaml config file. The first time after that when this misconfiguration will reveal itself is when flushing a memtable (see #15635), but it's good to know the endpoint is not configured earlier than that.

fixes: #15074

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4674/
- Duration: 6 hr 30 min
- Builder: monster

nyh commented on sstables/sstables_manager.hh:
```diff
@@ -71,6 +71,7 @@ public:
 
     storage_manager(const db::config&, config cfg);
     shared_ptr<s3::client> get_endpoint_client(sstring endpoint);
+    bool has_endpoint_client(sstring endpoint) const;
```
Nitpick: such a const function shouldn't take sstring by value (a copy), it's wasted effort - it should either take a `const sstring&`, or if the hash table implementation allows that, a std::string_view. But I'm mentioning this as a nitpick and not an important thing to fix because I see that the pre-existing get_endpoint_client() function has the same issue.

xemul replied:
Agree, `std::string_view` is much better. I'll fix them both in a followup.

nyh commented on test/object_store/test_basic.py:
```diff
@@ -135,6 +135,11 @@ async def test_basic(test_tempdir, s3_server, ssl):
+        desc = conn.execute(f"DESCRIBE KEYSPACE {ks}").one().create_statement
+        # The storage_opts wraps options with '{ <options> }' while the DESCRIBE
+        # does it like '{<options>}' so strip the corner brances and spaces for check
+        assert storage_opts.strip("{ }") in desc, "DESCRIBE generates unexpected storage options"
```
I'm in favor of moving this test from cql-pytest/ to object_store/, but I'm less happy about the fact that in object_store/ everything is tested in one big test function `test_basic()`, and here you make it even bigger. I think the following approach can be better:
1. Create a fixture that sets up a table on s3.
2. Create a small test function that **only** checks DESCRIBE KEYSPACE

Anyway - this is for your consideration. I don't insist. There's nothing really incorrect with your approach either.

xemul replied:
I like the making the table as "fixture" suggestion. Meanwhile there's less pytest-ish generalization of this place in #15986 and the idea to keep the whole object_store test in dtests (#16006). So I'd prefer to postpone this particular change for later.

xemul replied:
I remembered. Currently object_store test cases do not use clusters provided by test.py, they create it on their own explicitly inside the test case function. Making ks.cf a fixture would require making scyllas be fixtures too, so it's much more work. Also I don't remember why @tchaikov didn't make it a fixture from the very beginning, there was some reason.

tchaikov replied:
the reason is i am lazy. back then, i was trying to migrate the test to pytest and focused on transforming the cluster (scylla instance) and s3 server into fixtures . they are the facilities provided by `test.py` . but put much less efforts on restructuring the test itself.

and yeah, i agree that the table creation/teardown is a great candidate of fixturization.

xemul replied:
> the reason is i am lazy

:rofl:

nyh reviewed (changes requested):
I left some nitpicks for your considerations but I don't insist on them.

However, one thing I would like to insist on (unless you can convince me otherwise) is to please add at the end a test case that reproduces the actual bug you are fixing - i.e., that if you attempt to create a table with an unconfigured s3 endpoint, you get an error. We should start getting used to writing regression tests for bugs that we fix - i.e., a test that reproduces the bug before your patch and passes afterwards. I think it should be pretty easy to write such a test - in fact, if you want (though I'm not sure it's a good idea) you can even just leave the test in cql-pytest which you deleted in your first patch - and just change the test to check the creation fails.

When you change or write such a test, please add to it a comment that it reproduces the issue number whatever-the-number-is. Such a comment can be useful in the future if we notice the test failing on some branch and wonder what patch we forgot to backport.

xemul commented:
upd:
- added a test case to validate the original issue is fixed
- moved validation checks to database::validate_new_keyspace() (previous place was "too late already")

nyh commented on replica/database.cc:
```diff
@@ -1422,6 +1423,15 @@ void database::validate_new_keyspace(keyspace_metadata& ksm) {
+                throw exceptions::invalid_request_exception("Keyspace storage options not supported in the cluster");
+            }
+        }
+    }, ksm.get_storage_options().value);
```
The function name validate_new_keyspace() suggests (?) it only applies to CREATE KEYSPACE, not to ALTER KEYSPACE. Can you please add to your test also a check what happens if you try to ALTER a keyspace to use S3? Is this even possible - even if S3 *is* configured?

xemul replied:
> Can you please add to your test also a check what happens if you try to ALTER a keyspace to use S3? 

Changing keyspace storage type is not allowed:

```
# Test that storage options cannot be altered (at least until it's well defined
# what it means to e.g. switch from S3 to another format and back).
def test_storage_options_alter_type(cql, scylla_only):
    ksdef = "WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : '1' } " \
            "AND STORAGE = { 'type' : 'LOCAL' }"
    with new_test_keyspace(cql, ksdef) as keyspace:
         # It's not fine to change the storage type
        ksdef_local = "WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : '1' } " \
            "AND STORAGE = { 'type' : 'S3', 'bucket' : '/b1', 'endpoint': 'localhost'}"
        with pytest.raises(InvalidRequest):
            res = cql.execute(f"ALTER KEYSPACE {keyspace} {ksdef_local}")
```

nyh replied:
Thanks.

nyh commented on test/object_store/test_basic.py:
```diff
@@ -233,3 +234,20 @@ def list_bucket(s3_server):
+                                          'replication_factor': '1'})
+        storage_opts = format_tuples(type='S3',
+                                     endpoint='unknown_endpoint',
+                                     bucket=s3_server.bucket_name)
```
Why do you need the s3_server fixture, or its bucket_name, for this test? I think you don't need it?

tchaikov replied:
i concur with you. but i think Pavel might want to create a more real-world use case, a cluster is configured with some (correct) backend storage endpoints. but the user somehow tries to create a keyspace with a wrong endpoint.

xemul replied:
Well... I need a bucket name. It can be random as well, but strictly speaking the check is for misconfigured endpoint, not endpoint:bucket pair

nyh replied:
@tchaikov imagine that you are testing the "curl" tool and want to check what happens when you give it a malformed URL. Would you begin the test by running a real Web server on AWS and then running a curl command with a URL that has nothing to do with that server.

@xemul yes, you need a bucket name, but how is the bucket name from the real S3 server which you don't use relevant, more than the bucket name "dog"?

I admit it's not a disaster, but I just found it weird.

By the way, you keep saying that this is an "unconfigured" endpoint, but in fact, in this test you did configure an endpoint, you just set it to 'unknown_endpoint', which presumably doesn't exist in the DNS (but who knows, might ;-)). Shouldn't you also try not setting "endpoint" at all instead of setting it to garbage?

Another by the way - what happens if "endpoint" is it to an address that looks fine, e.g., "www.google.com", but it's just not an S3 server? Is this tested on creation time? I assume the answer is no?

xemul replied:
> By the way, you keep saying that this is an "unconfigured" endpoint, but in fact, in this test you did configure an endpoint, you just set it to 'unknown_endpoint', which presumably doesn't exist in the DNS (but who knows, might ;-)). Shouldn't you also try not setting "endpoint" at all instead of setting it to garbage?

Not setting endpoint is not allowed and there's test for that in cql-pytest/test_keyspace.py

```
# Test that not passing required parameters fails
def test_storage_options_required_param(cql, scylla_only):
    ksdef = "WITH REPLICATION = { 'class' : 'NetworkTopologyStrategy', 'replication_factor' : '1' } " \
            "AND STORAGE = { 'type' : 'S3', 'bucket' : '42' }"
    with pytest.raises(InvalidRequest):
        with new_test_keyspace(cql, ksdef):
            pass
```

> Another by the way - what happens if "endpoint" is it to an address that looks fine, e.g., "www.google.com", but it's just not an S3 server? Is this tested on creation time? I assume the answer is no?

Then it will create a client for that endpoint later and will get some error. Most likely this will first happen memtable flush time, for that case there's #15635

tchaikov replied:
> @tchaikov imagine that you are testing the "curl" tool and want to check what happens when you give it a malformed URL. Would you begin the test by running a real Web server on AWS and then running a curl command with a URL that has nothing to do with that server.
> 

ahh! good point. i am convinced. thanks.

nyh reviewed (changes requested):
Looks mostly good, I just left some tiny comments. Thanks.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4686/
- Duration: 2 hr 14 min
- Builder: spider1.cloudius-systems.com

tchaikov commented on replica/database.cc:
```diff
@@ -1430,6 +1430,10 @@ void database::validate_new_keyspace(keyspace_metadata& ksm) {
                 throw exceptions::invalid_request_exception("Keyspace storage options not supported in the cluster");
             }
+            // It's non-system keyspace
+            if (!_user_sstables_manager->has_endpoint_client(so.endpoint)) {
```
instead of requiring `database` to validate `storage_options`, probably it might be better if we can leave this to `sstable_manager` ?

it is already an "acquaintance" of `storage_options`, as it creates a new instance of `shared_sstable` with the specified `storage_option`.

xemul replied:
The point is to check endpoint presence when a keyspace is created, not when an sstable is created. The latter is _much_ later -- after a keyspace one would need to create a table, populate it with data and flush.

tchaikov replied:
yeah, i understand when this is called. i suggest to add a function in `sstable_manager` , and call that method instead. would be better to keep the `database` away from the internals of storage option verification.

tchaikov commented on sstables/sstables_manager.hh:
```diff
@@ -128,6 +129,11 @@ public:
         return _storage->get_endpoint_client(std::move(endpoint));
     }
 
+    bool has_endpoint_client(sstring endpoint) const {
```
IMHO, this name is kind of confusing. we are not checking if we have one or more clients for a given endpoint, instead we check if the endpoint is configured.

probably we can rename this method to `is_known_endpoint()`?

xemul replied:
> IMHO, this name is kind of confusing. we are not checking if we have one or more clients for a given endpoint, instead we check if the endpoint is configured.

Isn't it the same? In this particular place I mean

xemul replied:
Renamed anyway

xemul commented:
upd:
- rebased (has conflicst)
- rename has_endpoint_client -> is_known_endpoint

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4731/
- Duration: 7 hr 43 min
- Builder: i-077c0a37bd021f8d8

xemul commented:
upd:
- move storage options validation even deeper than databse -- into sstables manager

tchaikov reviewed (approved):
lgtm. thank you !

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4775/
- Duration: 6 hr 7 min
- Builder: i-0b0548df7579ed7cb
